aws-lc-rs/src/kem.rs

```diff
@@ -0,0 +1,504 @@
+            panic!();
+        };
+
+        Ok(EncapsulationKey {
+            algorithm: self.algorithm,
+            evp_pkey,
+        })
+    }
+
+    /// Performs the decapsulate operation using this KEM decapsulation key on the given ciphertext.
+    ///
+    /// `ciphertext` is the ciphertext generated by the encapsulate operation using the KEM encapsulation key
+    /// associated with this KEM decapsulation key.
+    ///
+    /// # Errors
+    /// `Unspecified` when operation fails due to internal error.
+    #[allow(clippy::needless_pass_by_value)]
+    pub fn decapsulate(&self, ciphertext: Ciphertext<'_>) -> Result<SharedSecret, Unspecified> {
+        let mut shared_secret_len = self.algorithm.shared_secret_size();
+        let mut shared_secret: Vec<u8> = vec![0u8; shared_secret_len];
+
+        let ctx = LcPtr::new(unsafe { EVP_PKEY_CTX_new(*self.evp_pkey, null_mut()) });
+        let ctx = if let Ok(ctx) = ctx {
+            ctx
+        } else {
+            return Err(Unspecified);
+        };
+
+        let ciphertext = ciphertext.as_ref();
+
+        if 1 != unsafe {
+            EVP_PKEY_decapsulate(
+                *ctx,
+                shared_secret.as_mut_ptr(),
+                &mut shared_secret_len,
+                // AWS-LC incorrectly has this as an unqualified `uint8_t *`, it should be qualified with const
+                ciphertext.as_ptr() as *mut u8,
+                ciphertext.len(),
+            )
+        } {
+            return Err(Unspecified);
+        }
+
+        shared_secret.truncate(shared_secret_len);
+
+        Ok(SharedSecret(shared_secret.into_boxed_slice()))
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn new(alg: &'static Algorithm<Id>, bytes: &[u8]) -> Result<Self, KeyRejected> {
+        match bytes.len().cmp(&alg.decapsulate_key_size()) {
+            Ordering::Less => Err(KeyRejected::too_small()),
+            Ordering::Greater => Err(KeyRejected::too_large()),
+            Ordering::Equal => Ok(()),
+        }?;
+        let privkey = LcPtr::new(unsafe {
+            EVP_PKEY_kem_new_raw_secret_key(alg.id.nid(), bytes.as_ptr(), bytes.len())
+        })?;
+        Ok(DecapsulationKey {
+            algorithm: alg,
+            evp_pkey: privkey,
+        })
+    }
+}
+
+unsafe impl<Id> Send for DecapsulationKey<Id> where Id: AlgorithmIdentifier {}
+
+unsafe impl<Id> Sync for DecapsulationKey<Id> where Id: AlgorithmIdentifier {}
+
+impl<Id> Debug for DecapsulationKey<Id>
+where
+    Id: AlgorithmIdentifier,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("DecapsulationKey")
+            .field("algorithm", &self.algorithm)
+            .finish_non_exhaustive()
+    }
+}
+
+mod encoding {
+    pub struct PublicKeyBinDerType {
+        _priv: (),
+    }
+}
+
+/// A KEM Encapsulation Key Big-Endian number representation encoded using DER.
+pub type EncapsulationKeyBinDer = Buffer<'static, encoding::PublicKeyBinDerType>;
```

Comment:
what is this? The encapsulation key is not necessarily a number, for example in Kyber.

Reply:
Alright so what is it? We are trying to have a type that signifies that the given EncapsulationKey is being serialized to some sort of DER encoded value. Do we just call it `EncapsulationKeyRawDer`?

Reply:
> Alright so what is it?

it can be different things, depends on the KEM.

> Do we just call it EncapsulationKeyRawDer?

sounds good to me.

Reply:
Hmm I double checked and believe there was a misunderstanding on my part that this was encoded as ASN.1 as well, when it was not. Further investigation appears that these bytes, in the case of the draft ML-KEM specification, is actually little-endian encoded? I assume this is another aspect for KEM that could become algorithm dependent. 

Since this type will be generic over multiple algorithms that match the KEM interface, I wonder if we should just support raw public keys using the `AsRef<[u8]>` trait instead? This would seem more appropriate then having trait implementations of `AsBigEndian` or `AsLittleEndian` as those will be algorithm dependent, and we should only support the representations that are inline with the specifications. We'd then wait to have a standardized serialization format like PKCS#8 / X.509 SubjectPublicKeyInfo before implementing an encoding trait.

Otherwise if we don't want this to behind a flexible interface, then maybe we should take a step-back and adopt a design that is closer to the [AES Key Wrap](https://github.com/aws/aws-lc-rs/pull/250) module design. Where we have a general trait on the behavior, and implementations of the behavior as child modules.

Thoughts @justsmth & @dkostic ?

Reply:
> Since this type will be generic over multiple algorithms that match the KEM interface, I wonder if we should just support raw public keys using the AsRef<[u8]> trait instead?

I think that's a good idea. Once a serialization format is standardized we can define a proper encoding.